seasonal_bulletin.Rmd

```diff
@@ -0,0 +1,244 @@
+
+if (SEASON == "MAM") {
+  months_to_include <- c(3, 4, 5)
+} else if (SEASON == "OND") {
+  months_to_include <- c(10, 11, 12)
+}
+
+# Only relevant for SEAS5 data
+lead_times <- (months_to_include + 12 * (months_to_include < issued_month)) - issued_month
+
+# For Ethiopia, we're looking at PiN data
+if (ISO3 == "ETH") {
+  gdf_pop <- get_eth_gdf_pin(YEAR) # This contains all adms in ETH
+  if (ADM_LEVEL == 2) {
+    sel_zones <- subset_adm2_ond_mam()
+  } else {
+    sel_zones <- NULL
+  }
+  pcode_col <- glue("admin{ADM_LEVEL}Pcode")
+  pop_var <- "PiN"
+
+# While for Somalia we're looking at population
+} else if (ISO3 == "SOM") {
+  gdf_pop <- get_som_gdf_pop(ADM_LEVEL)
+  sel_zones <- NULL
+  pcode_col <- glue("ADM{ADM_LEVEL}_PCODE")
+  pop_var <- "Population"
+}
+
+```
+
+First load in the historical rainfall data from the Postgres database. We'll process that data to filter to only the months of interest (depending on the season we're looking at), 
+and calculate the total precipitation for that season, per pcode and year. We'll then calculate return periods and label tercile/quartiles for each pcode and year. 
+
+```{r loading}
+
+df_rainfall <- get_historical_rainfall(DATASET, ISO3, ADM_LEVEL, sel_zones)
+
+if (DATASET == "seas5") {
+  df_rainfall <- process_seas5_rainfall(
+    df_rainfall,
+    sel_month = issued_month,
+    months_to_include = months_to_include
+  )
+} else if (DATASET == "era5") {
+  df_rainfall <- process_era5_rainfall(
+    df_rainfall,
+    months_to_include = months_to_include
+  )
+}
+
+df_summary <- join_pop_rainfall(gdf_pop, df_rainfall, pcode_col)
+```
+
+``` {r load_cog_stack}
+
+containers <- cumulus::blob_containers(stage="prod")
+dir_name <- paste0(DATASET, "/monthly/processed")
+
+cog_df <- AzureStor::list_blobs(
+  container = containers$raster,
+  dir = dir_name
+)
+
+# process cog_df slightly differently for seas5 or era5
+if (DATASET == "seas5") {
+  cog_df <- cog_df %>%
+    mutate(
+      year = str_extract(name, "i(\\d{4})") %>% str_remove("i") %>% as.integer(), 
+      month = str_extract(name, "-(\\d{2})-") %>% str_remove_all("-") %>% as.integer(),
+      lt = str_extract(name, "lt\\d+") %>% str_remove("lt") %>% as.integer()
+    ) %>%
+    filter(
+      month == issued_month,
+      lt %in% lead_times
+    )
+  
+} else if (DATASET == "era5") {
+  cog_df <- cog_df %>%
+    mutate(
+      year = str_extract(name, "v(\\d{4})") %>% str_remove("v") %>% as.integer(),
+      month = str_extract(name, "-(\\d{2})-") %>% str_remove_all("-") %>% as.integer()
+    ) %>%
+    filter(month %in% months_to_include)
+}
+
+# Load by year, take the sum, then stack
+yearly_sums <- list()
+for (sel_year in unique(cog_df$year)) {
+  yearly_cogs <- filter(cog_df, year == sel_year)
+  urls <- paste0("/vsiaz/raster/", yearly_cogs$name)
+  cogs <- rast(urls)
+  cogs_clipped <- crop(cogs, gdf_pop)
+  cogs_masked <- mask(cogs_clipped, gdf_pop)
+  # go from mm/day to mm/month
+  yearly_sum <- sum((cogs_masked * 30))
```

Comment:
AC - to be more correct you would extract the date from the band names and use the lubridate::days_in_month() function again

Reply:
That function is so nice!!